File: oltrack-py/src/oltrack/adapters/sf.py
"""Adapter for 顺丰速运 / SF Express Open Platform (api.sf-express.com).

Reference response shape from mappings/cn/sf.csv source notes:

    {
      "apiResultCode": "A0000",
      "msgData": {
        "waybillNo": "SF1234567890",
        "expressStatus": 6,
        "expressStatusDesc": "已签收",
        "originCode": "...",
        "destCode": "...",
        "routeResps": [
          {
            "acceptTime": "2024-06-04 18:42:11",
            "acceptAddress": "北京朝阳CBD营业部",
            "remark": "客户签收",
            "opcode": "160",
            "opcodeDesc": "客户签收",
            "operatorPhone": "139****5678",
            "routeFrom": "北京朝阳CBD营业部",
            "routeTo": ""
          },
          ...
        ]
      }
    }

The adapter maps each routeResps[] item to a TrackingEvent dict, prefixing
opcode with "opcode:" to match the mapping CSV convention.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any
from zoneinfo import ZoneInfo

from oltrack import normalize

_SH_TZ = ZoneInfo("Asia/Shanghai")


def _isoformat_local(time_str: str) -> str:
    """Parse "2024-06-04 18:42:11" and attach Asia/Shanghai timezone."""
    naive = datetime.strptime(time_str, "%Y-%m-%d %H:%M:%S")
    return naive.replace(tzinfo=_SH_TZ).isoformat()


def normalize_event(route_resp: dict[str, Any]) -> dict[str, Any]:
    """Convert one routeResps[] item to an OLTS TrackingEvent."""
    raw_code = f"opcode:{route_resp['opcode']}"
    location = {}
    if addr := route_resp.get("acceptAddress"):
        location["name"] = addr
    metadata: dict[str, Any] = {}
    if route_from := route_resp.get("routeFrom"):
        metadata["routeFrom"] = route_from
    if route_to := route_resp.get("routeTo"):
        metadata["routeTo"] = route_to

    operator = None
    if phone := route_resp.get("operatorPhone"):
        operator = {"phone": phone}

    return normalize.event(
        carrier="sf",
        carrier_event_code=raw_code,
        event_time=_isoformat_local(route_resp["acceptTime"]),
        description=route_resp.get("remark") or route_resp.get("opcodeDesc"),
        location=location or None,
        operator=operator,
        metadata=metadata or None,
    )


def normalize_response(raw: dict[str, Any]) -> list[dict[str, Any]]:
    """Parse a full SF response into a list of TrackingEvent dicts."""
    if raw.get("apiResultCode") != "A0000":
        raise ValueError(f"SF API error: {raw.get('apiErrorMsg', raw.get('apiResultCode'))}")
    msg = raw.get("msgData", {})
    routes = msg.get("routeResps", [])
    return [normalize_event(r) for r in routes]


def normalize_to_shipment(raw: dict[str, Any]) -> dict[str, Any]:
    """Parse a full SF response into an OLTS Shipment dict."""
    msg = raw.get("msgData", {})
    events = normalize_response(raw)
    return normalize.shipment(
        tracking_number=msg["waybillNo"],
        events=events,
        carrier="sf",
        metadata={"expressStatus": msg.get("expressStatus")} if msg.get("expressStatus") else None,
    )

File: oltrack-ts/src/adapters/sf.ts
import type { TrackingEvent, UlscCode } from "../types.js";
import { event, shipment } from "../normalize.js";

/**
 * Subset of SF opcode → UlscCode mapping (from mappings/cn/sf.csv).
 *
 * In v0.5 MVP we inline a partial map for the most common codes. v1.0 will
 * auto-generate the full table from mappings/cn/sf.csv at build time.
 */
const SF_OPCODE_TO_ULSC: Record<string, UlscCode> = {
  "10": "order_created",
  "30": "picked_up",
  "50": "in_transit",
  "60": "arrived_at_hub",
  "70": "departed_from_hub",
  "80": "arrived_at_destination",
  "100": "out_for_delivery",
  "160": "delivered",
  "161": "signed_by_third_party",
  "170": "refused",
};

interface SfRouteResp {
  acceptTime: string;
  acceptAddress?: string;
  remark?: string;
  opcode: string;
  opcodeDesc?: string;
  operatorPhone?: string;
  routeFrom?: string;
  routeTo?: string;
  latitude?: string;
  longitude?: string;
}

interface SfResponse {
  apiResultCode?: string;
  apiErrorMsg?: string;
  msgData?: {
    waybillNo?: string;
    expressStatus?: number;
    routeResps?: SfRouteResp[];
  };
}

const TZ_OFFSET = "+08:00"; // Asia/Shanghai

function parseSfTime(s: string): string {
  // "2024-06-04 18:42:11" → "2024-06-04T18:42:11+08:00"
  return `${s.replace(" ", "T")}${TZ_OFFSET}`;
}

export function normalizeEvent(route: SfRouteResp): TrackingEvent {
  const code = SF_OPCODE_TO_ULSC[route.opcode];
  if (!code) {
    throw new Error(
      `unknown SF opcode ${route.opcode}; extend SF_OPCODE_TO_ULSC or fall back to full mappings/cn/sf.csv lookup`
    );
  }
  const carrierEventCode = `opcode:${route.opcode}`;
  const location = route.acceptAddress ? { name: route.acceptAddress } : undefined;
  const operator = route.operatorPhone ? { phone: route.operatorPhone } : undefined;
  const metadata: Record<string, unknown> = {};
  if (route.routeFrom) metadata.routeFrom = route.routeFrom;
  if (route.routeTo) metadata.routeTo = route.routeTo;

  return event({
    carrierCode: "sf",
    carrierEventCode,
    eventTime: parseSfTime(route.acceptTime),
    oltsCode: code,
    description: route.remark ?? route.opcodeDesc,
    location,
    operator,
    metadata: Object.keys(metadata).length > 0 ? metadata : undefined,
  });
}

export function normalizeResponse(raw: SfResponse): TrackingEvent[] {
  if (raw.apiResultCode !== "A0000") {
    throw new Error(`SF API error: ${raw.apiErrorMsg ?? raw.apiResultCode}`);
  }
  const routes = raw.msgData?.routeResps ?? [];
  return routes.map(normalizeEvent);
}

export function normalizeToShipment(raw: SfResponse) {
  const msg = raw.msgData ?? {};
  if (!msg.waybillNo) throw new Error("SF response missing waybillNo");
  const events = normalizeResponse(raw);
  return shipment({
    trackingNumber: msg.waybillNo,
    carrierCode: "sf",
    events,
    metadata: msg.expressStatus !== undefined ? { expressStatus: msg.expressStatus } : undefined,
  });
}
